cedar-policy-core/src/authorizer/partial_response.rs

```diff
@@ -0,0 +1,481 @@
+        let nontrivial = self
+            .residual_permits
+            .iter()
+            .map(|(id, (r, a))| (id, (r, a)));
+        trues.chain(falses).chain(nontrivial)
+    }
+
+    /// Returns all residuals expressions that come from [`Effect::Forbid`] policies
+    fn all_forbid_residuals(&self) -> impl Iterator<Item = (&PolicyID, (&Expr, &Annotations))> {
+        let trues = self
+            .satisfied_forbids
+            .iter()
+            .map(|(id, a)| (id, (&self.true_expr, a)));
+        let falses = self
+            .false_forbids
+            .iter()
+            .map(|(id, (_, a))| (id, (&self.false_expr, a)));
+        let nontrivial = self
+            .residual_forbids
+            .iter()
+            .map(|(id, (r, a))| (id, (r, a)));
+        trues.chain(falses).chain(nontrivial)
+    }
+
+    /// Return the residual for a given [`PolicyID`], if it exists in the response
+    pub fn get(&self, id: &PolicyID) -> Option<&Expr> {
+        self.get_true(id)
+            .or_else(|| self.get_false(id).or_else(|| self.get_residual(id)))
+    }
+
+    fn get_true(&self, id: &PolicyID) -> Option<&Expr> {
+        self.satisfied_permits
+            .get(id)
+            .or_else(|| self.satisfied_forbids.get(id))
+            .map(|_| &self.true_expr)
+    }
+
+    fn get_false(&self, id: &PolicyID) -> Option<&Expr> {
+        self.false_permits
+            .get(id)
+            .or_else(|| self.false_forbids.get(id))
+            .map(|_| &self.false_expr)
+    }
+
+    fn get_residual(&self, id: &PolicyID) -> Option<&Expr> {
+        self.residual_permits
+            .get(id)
+            .or_else(|| self.residual_forbids.get(id))
+            .map(|(r, _)| r)
+    }
+
+    /// Attempt to re-authorize this response given a mapping from unknowns to values
+    pub fn reauthorize(
+        &self,
+        mapping: &HashMap<SmolStr, Value>,
+        auth: &Authorizer,
+        r: Request,
+        es: &Entities,
+    ) -> Result<Self, PolicySetError> {
+        let policyset = self.all_policies(mapping)?;
+        Ok(auth.is_authorized_core(r, &policyset, es))
+    }
+
+    fn all_policies<'a>(
+        &'a self,
+        mapping: &'a HashMap<SmolStr, Value>,
+    ) -> Result<PolicySet, PolicySetError> {
+        PolicySet::try_from_iter(
+            build_policies(Effect::Permit, self.all_permit_residuals(), mapping).chain(
+                build_policies(Effect::Forbid, self.all_forbid_residuals(), mapping),
+            ),
+        )
+    }
+
+    fn errors(self) -> impl Iterator<Item = AuthorizationError> {
+        self.residual_forbids
+            .into_iter()
+            .chain(self.residual_permits)
+            .map(
+                |(id, (expr, _))| AuthorizationError::PolicyEvaluationError {
+                    id: id.clone(),
+                    error: EvaluationError::non_value(expr),
+                },
+            )
+            .chain(self.errors)
```

Comment:
for discussion: why does this add `NonValue` errors for each of the nontrivial residuals, as opposed to just returning `self.errors` only?

Reply:
Trying to be inline with the regular cedar error policy. Policies that don't evaluate to a value are an error.